thread_sync.c

```diff
@@ -1029,13 +1029,19 @@ static VALUE
 queue_do_pop(VALUE self, struct rb_queue *q, int should_block, VALUE timeout)
 {
     check_array(self, q->que);
-    rb_hrtime_t end = queue_timeout2hrtime(timeout);
+    if (RARRAY_LEN(q->que) == 0) {
+      if (!should_block) {
+          rb_raise(rb_eThreadError, "queue empty");
+      }
+
+      if (RTEST(rb_equal(INT2FIX(0), timeout))) {
```

Comment:
@eregon this should do ^

Reply:
That should work, except if it's a custom object with `to_f`. Probably good enough.